pyat/at/collective/beam_loading.py

```diff
@@ -0,0 +1,282 @@
+    new_elems = []
+    for ref, qf, rs in zip(cavpts, qfactor, rshunt):
+        cav = ring[ref]
+        assert isinstance(cav, RFCavity), \
+            'Beam loading can only be assigned to a cavity element'
+        new_elems.append(BeamLoadingElement.build_from_cav(cav, ring, qf,
+                                                           rs, **kwargs))
+    return apply(ring, cavpts, new_elems)
+
+
+def remove_beamloading(ring, cavpts: Refpts = None,
+                       copy: Optional[bool] = False):
+    """Function to remove beam loading from a cavity element, the beam
+    loading element is changed to a beam loading element that
+    combines the energy kick from both the cavity and the resonator
+
+    Parameters:
+        ring:           Lattice object
+
+    Keyword Arguments:
+        cavpts (Refpts):    refpts of the beam loading Elements.
+                            If None (default) apply to all elements
+        copy:       If True, returns a shallow copy of ring with new
+                      cavity elements. Otherwise, modify ring in-place
+    """
+    @make_copy(copy)
+    def apply(ring, cavpts, newelems):
+        for ref, elem in zip(cavpts, newelems):
+            ring[ref] = elem
+
+    if cavpts is None:
+        cavpts = ring.get_refpts(BeamLoadingElement)
+    else:
+        cavpts = uint32_refpts(cavpts, len(ring))
+    new_elems = []
+    for ref in cavpts:
+        bl = ring[ref]
+        assert isinstance(bl, BeamLoadingElement), \
+            'Cannot remove beam loading: not a beam loading element'
+        family_name = bl.FamName.replace('_BL', '')
+        harm = numpy.round(bl.Frequency*ring.circumference/clight)
+        new_elems.append(RFCavity(family_name, bl.Length, bl.Voltage,
+                                  bl.Frequency, harm, bl.Energy,
+                                  TimeLag=getattr(bl, 'TimeLag', 0.0),
+                                  PhaseLag=getattr(bl, 'PhaseLag', 0.0)))
+    return apply(ring, cavpts, new_elems)
+
+
+class BeamLoadingElement(RFCavity, Collective):
+    """Class to generate a beamloading element, inherits from Element
+       additional argument are ring, cavity, qfactor, rshunt
+    """
+    default_pass = {False: 'DriftPass', True: 'BeamLoadingCavityPass'}
+    _conversions = dict(RFCavity._conversions,
+                        Rshunt=float, Qfactor=float, NormFact=float,
+                        PhaseGain=float, VoltGain=float, _mode=int,
+                        _beta=float, _wakefact=float, _nslice=int,
+                        ZCuts=lambda v: _array(v),
+                        _nturns=int, _phis=float,
+                        _turnhistory=lambda v: _array(v),
+                        _vbunch=lambda v: _array(v),
+                        _vbeam_phasor=lambda v: _array(v),
+                        _vbeam=lambda v: _array(v),
+                        _vcav=lambda v: _array(v),
+                        _vgen=lambda v: _array(v))
```

Comment:
Apparently, `_vgen` must have shape (2,) and can be set through `@Vgen.setter`. You could enforce the right shape with:
```python
_vgen=lambda v: _array(v, shape=(2,)))
```
It's also true for other attributes, but if they cannot be set by the user, it's not so important.